CSharp Client Generator - Not returning result nor throwing exception for non-success HTTP status codes if IsPlainText = true

Hi, 

I'm having an issue trying generate a C# client out of an OpenAPI spec which contains plain/text responses.

Specifically, I have this error telling me that **not all code paths return a value** if the operations have **non-success HTTP status codes** as you can see below:
![image](https://user-images.githubusercontent.com/66265246/124018219-007ae600-d9be-11eb-88d6-fe6f1d0caafd.png)

I think that the problem happens here:
https://github.com/RicoSuter/NSwag/blob/1c836ee3cc766ce25f3170618a5d12e5e96fde61/src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.ProcessResponse.liquid#L15

Basically we will need to **add an else block** to deal with that non-success HTTP codes and throw an exception properly.

Does that make sense? If it does I can fix this and raise a PR.
Thanks!

Hi.

I tried to use the updated template file `Client.Class.ProcessResponse.liquid` but I get the following compiler error:

```
error CS0103: The name 'objectResponse_' does not exist in the current context
```

I'm using the `NSwag.ApiDescription.Client` nuget package version 13.11.3.

```
    <OpenApiReference Include="ApiClient.json" CodeGenerator="NSwagCSharp" Namespace="MyNamespace.ApiClient" ClassName="ApiClient" Options="/ClientBaseClass:ApiClientBase /UseHttpRequestMessageCreationMethod:true /TemplateDirectory:template">
      <SourceUri>http://localhost/MyWebApplication/docs/v1/swagger.json</SourceUri>
      <OutputPath>..\ApiClient.cs</OutputPath>
    </OpenApiReference>
```

Any ideas?

Thank you.

--- a/src/NSwag.CodeGeneration.CSharp.Tests/PlainResponseTests.cs
+++ b/src/NSwag.CodeGeneration.CSharp.Tests/PlainResponseTests.cs
@@ -61,14 +61,14 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 }";
             var document = await OpenApiDocument.FromJsonAsync(json);
 
-            //// Act
+            // Act
             var codeGenerator = new CSharpClientGenerator(document, new CSharpClientGeneratorSettings
             {
                 GenerateClientInterfaces = true
             });
             var code = codeGenerator.GenerateFile();
 
-            //// Assert
+            // Assert
             Assert.Contains("public async System.Threading.Tasks.Task<System.Collections.Generic.ICollection<string>> GetIenumerableAsync(", code);
             Assert.Contains("await ReadObjectResponseAsync<System.Collections.Generic.ICollection<string>>(", code);
         }
@@ -103,14 +103,14 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 }";
             var document = await OpenApiDocument.FromJsonAsync(json);
 
-            //// Act
+            // Act
             var codeGenerator = new CSharpClientGenerator(document, new CSharpClientGeneratorSettings
             {
                 GenerateClientInterfaces = true
             });
             var code = codeGenerator.GenerateFile();
 
-            //// Assert
+            // Assert
             Assert.Contains("public async System.Threading.Tasks.Task<string> PlainAsync(", code);
             Assert.Contains("(string)System.Convert.ChangeType(responseData_, typeof(string));", code);
         }
@@ -146,14 +146,14 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 }";
             var document = await OpenApiDocument.FromJsonAsync(swagger);
 
-            //// Act
+            // Act
             var codeGenerator = new CSharpClientGenerator(document, new CSharpClientGeneratorSettings
             {
                 GenerateClientInterfaces = true
             });
             var code = codeGenerator.GenerateFile();
 
-            //// Assert
+            // Assert
             Assert.Contains("public async System.Threading.Tasks.Task<System.Collections.Generic.ICollection<string>> GetIenumerableAsync(", code);
             Assert.Contains("await ReadObjectResponseAsync<System.Collections.Generic.ICollection<string>>(", code);
         }
@@ -184,16 +184,15 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 }";
             var document = await OpenApiDocument.FromJsonAsync(swagger);
 
-            //// Act
+            // Act
             var codeGenerator = new CSharpClientGenerator(document, new CSharpClientGeneratorSettings
             {
                 GenerateClientInterfaces = true
             });
             var code = codeGenerator.GenerateFile();
 
-            //// Assert
+            // Assert
             Assert.Contains("public async System.Threading.Tasks.Task<string> PlainAsync(", code);
-            Assert.Contains("(string)System.Convert.ChangeType(responseData_, typeof(string));", code);
         }
     }
 }

--- a/src/NSwag.CodeGeneration.CSharp.Tests/PlainResponseTests.cs
+++ b/src/NSwag.CodeGeneration.CSharp.Tests/PlainResponseTests.cs
@@ -10,6 +10,70 @@ namespace NSwag.CodeGeneration.CSharp.Tests
     public class PlainResponseTests
     {
         [Fact]
+        public async Task When_openapi3_reponse_contains_plain_text_string_array_then_ReadObjectResponseAsync_is_generated()
+        {
+            // Arrange
+            var json = @"{
+  ""openapi"": ""3.0.1"",
+  ""paths"": {
+    ""/api/get-ienumerable"": {
+      ""get"": {
+        ""responses"": {
+          ""200"": {
+            ""description"": ""Success"",
+            ""content"": {
+              ""text/plain"": {
+                ""schema"": {
+                  ""type"": ""array"",
+                  ""items"": {
+                    ""type"": ""string"",
+                    ""nullable"": true
+                  },
+                  ""nullable"": true
+                }
+              },
+              ""application/json"": {
+                ""schema"": {
+                  ""type"": ""array"",
+                  ""items"": {
+                    ""type"": ""string"",
+                    ""nullable"": true
+                  },
+                  ""nullable"": true
+                }
+              },
+              ""text/json"": {
+                ""schema"": {
+                  ""type"": ""array"",
+                  ""items"": {
+                    ""type"": ""string"",
+                    ""nullable"": true
+                  },
+                  ""nullable"": true
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}";
+            var document = await OpenApiDocument.FromJsonAsync(json);
+
+            //// Act
+            var codeGenerator = new CSharpClientGenerator(document, new CSharpClientGeneratorSettings
+            {
+                GenerateClientInterfaces = true
+            });
+            var code = codeGenerator.GenerateFile();
+
+            //// Assert
+            Assert.Contains("public async System.Threading.Tasks.Task<System.Collections.Generic.ICollection<string>> GetIenumerableAsync(", code);
+            Assert.Contains("await ReadObjectResponseAsync<System.Collections.Generic.ICollection<string>>(", code);
+        }
+
+        [Fact]
         public async Task When_openapi3_reponse_contains_plain_text_then_Convert_is_generated()
         {
             // Arrange
@@ -48,6 +112,50 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 
             //// Assert
             Assert.Contains("public async System.Threading.Tasks.Task<string> PlainAsync(", code);
+            Assert.Contains("(string)System.Convert.ChangeType(responseData_, typeof(string));", code);
+        }
+
+        [Fact]
+        public async Task When_swagger_reponse_contains_plain_text_string_array_then_ReadObjectResponseAsync_is_generated()
+        {
+            // Arrange
+            var swagger = @"{
+  ""swagger"": ""2.0"",
+  ""paths"": {
+    ""/api/get-ienumerable"": {
+      ""get"": {
+        ""produces"": [
+          ""text/plain"",
+          ""application/json"",
+          ""text/json""
+        ],
+        ""responses"": {
+          ""200"": {
+            ""description"": ""Success"",
+            ""schema"": {
+              ""type"": ""array"",
+              ""items"": {
+                ""type"": ""string""
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}";
+            var document = await OpenApiDocument.FromJsonAsync(swagger);
+
+            //// Act
+            var codeGenerator = new CSharpClientGenerator(document, new CSharpClientGeneratorSettings
+            {
+                GenerateClientInterfaces = true
+            });
+            var code = codeGenerator.GenerateFile();
+
+            //// Assert
+            Assert.Contains("public async System.Threading.Tasks.Task<System.Collections.Generic.ICollection<string>> GetIenumerableAsync(", code);
+            Assert.Contains("await ReadObjectResponseAsync<System.Collections.Generic.ICollection<string>>(", code);
         }
 
         [Fact]
@@ -85,6 +193,7 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 
             //// Assert
             Assert.Contains("public async System.Threading.Tasks.Task<string> PlainAsync(", code);
+            Assert.Contains("(string)System.Convert.ChangeType(responseData_, typeof(string));", code);
         }
     }
 }

--- a/src/NSwag.CodeGeneration.CSharp.Tests/ResponseTests.cs
+++ b/src/NSwag.CodeGeneration.CSharp.Tests/ResponseTests.cs
@@ -8,6 +8,7 @@ namespace NSwag.CodeGeneration.CSharp.Tests
         [Fact]
         public async Task When_response_references_object_in_inheritance_hierarchy_then_return_value_is_correct()
         {
+            // Arrange
             var json = @"{
   ""openapi"": ""3.0.0"",
   ""paths"": {
@@ -64,12 +65,12 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 }";
             var document = await OpenApiDocument.FromJsonAsync(json);
 
-            //// Act
+            // Act
             var settings = new CSharpClientGeneratorSettings { ClassName = "MyClass" };
             var generator = new CSharpClientGenerator(document, settings);
             var code = generator.GenerateFile();
 
-            //// Act
+            // Assert
             Assert.Contains("System.Threading.Tasks.Task<ValidationException>", code);
             Assert.DoesNotContain("System.Threading.Tasks.Task<object>", code);
             Assert.Contains("class BusinessException", code);
@@ -79,6 +80,7 @@ namespace NSwag.CodeGeneration.CSharp.Tests
         [Fact]
         public async Task When_response_is_referenced_any_then_class_is_generated()
         {
+            // Arrange
             var json = @"{
   ""openapi"": ""3.0.0"",
   ""paths"": {
@@ -110,12 +112,12 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 }";
             var document = await OpenApiDocument.FromJsonAsync(json);
 
-            //// Act
+            // Act
             var settings = new CSharpClientGeneratorSettings { ClassName = "MyClass" };
             var generator = new CSharpClientGenerator(document, settings);
             var code = generator.GenerateFile();
 
-            //// Act
+            // Assert
             Assert.Contains("System.Threading.Tasks.Task<BusinessException>", code);
             Assert.DoesNotContain("System.Threading.Tasks.Task<object>", code);
             Assert.Contains("class BusinessException", code);
@@ -124,6 +126,7 @@ namespace NSwag.CodeGeneration.CSharp.Tests
         [Fact]
         public async Task When_same_response_is_referenced_multiple_times_in_operation_then_class_is_generated()
         {
+            // Arrange
             string json = @"{
   ""openapi"": ""3.0.0"",
   ""paths"": {
@@ -168,12 +171,12 @@ namespace NSwag.CodeGeneration.CSharp.Tests
 
             var document = await OpenApiDocument.FromJsonAsync(json);
 
-            //// Act
+            // Act
             var settings = new CSharpClientGeneratorSettings { ClassName = "MyClass" };
             var generator = new CSharpClientGenerator(document, settings);
             var code = generator.GenerateFile();
 
-            //// Act
+            // Assert
             Assert.Contains("System.Threading.Tasks.Task<System.Collections.Generic.ICollection<BusinessException>>", code);
             Assert.Contains("class BusinessException", code);
         }
@@ -181,6 +184,7 @@ namespace NSwag.CodeGeneration.CSharp.Tests
         [Fact]
         public async Task When_responses_produce_multiple_types()
         {
+            // Arrange
             string json = @"{
   ""openapi"": ""3.0.1"",
   ""info"": {
@@ -344,10 +348,181 @@ namespace NSwag.CodeGeneration.CSharp.Tests
             var generator = new CSharpClientGenerator(document, settings);
             var code = generator.GenerateFile();
 
-            // Act
+            // Assert
             Assert.Contains("var result_ = (string)System.Convert.ChangeType(responseData_, typeof(string));", code);
             Assert.Contains("await ReadObjectResponseAsync<ValidationProblemDetails>", code);
             Assert.Contains("await ReadObjectResponseAsync<ProblemDetails>", code);
+        }
+
+        [Fact]
+        public async Task When_responses_produce_primitive_types()
+        {
+            // Arrange
+            string json = @"{
+  ""swagger"": ""2.0"",
+  ""info"": {
+    ""title"": ""My Title"",
+    ""version"": ""1.0.0""
+  },
+  ""host"": ""localhost:13452"",
+  ""schemes"": [
+    ""http""
+  ],
+  ""consumes"": [
+    ""application/json""
+  ],
+  ""produces"": [
+    ""application/json""
+  ],
+  ""paths"": {
+    ""/api/Geo/UploadFile"": {
+      ""post"": {
+        ""operationId"": ""Geo_UploadFile"",
+        ""consumes"": [
+          ""multipart/form-data""
+        ],
+        ""parameters"": [
+          {
+            ""type"": ""file"",
+            ""name"": ""file"",
+            ""in"": ""formData"",
+            ""x-nullable"": true
+          }
+        ],
+        ""responses"": {
+          ""200"": {
+            ""x-nullable"": false,
+            ""description"": """",
+            ""schema"": {
+              ""type"": ""boolean""
+            }
+          }
+        }
+      }
+    },
+    ""/api/Geo/PostDouble"": {
+      ""post"": {
+        ""operationId"": ""Geo_PostDouble"",
+        ""parameters"": [
+          {
+            ""type"": ""number"",
+            ""name"": ""value"",
+            ""in"": ""query"",
+            ""format"": ""double"",
+            ""x-nullable"": true
+          }
+        ],
+        ""responses"": {
+          ""200"": {
+            ""x-nullable"": true,
+            ""description"": """",
+            ""schema"": {
+              ""type"": ""number"",
+              ""format"": ""double""
+            }
+          }
+        }
+      }
+    },
+    ""/api/Persons/{id}/Name"": {
+      ""get"": {
+        ""operationId"": ""Persons_GetName"",
+        ""parameters"": [
+          {
+            ""type"": ""string"",
+            ""name"": ""id"",
+            ""in"": ""path"",
+            ""required"": true,
+            ""description"": ""The person ID."",
+            ""format"": ""guid"",
+            ""x-nullable"": false
+          }
+        ],
+        ""responses"": {
+          ""200"": {
+            ""x-nullable"": false,
+            ""schema"": {
+              ""type"": ""string""
+            }
+          }
+        }
+      }
+    },
+    ""/api/Persons/AddXml"": {
+      ""post"": {
+        ""operationId"": ""Persons_AddXml"",
+        ""consumes"": [
+          ""application/xml""
+        ],
+        ""parameters"": [
+          {
+            ""name"": ""person"",
+            ""in"": ""body"",
+            ""schema"": {
+              ""type"": ""string"",
+              ""x-nullable"": true
+            },
+            ""x-nullable"": true
+          }
+        ],
+        ""responses"": {
+          ""200"": {
+            ""x-nullable"": true,
+            ""description"": """",
+            ""schema"": {
+              ""type"": ""string""
+            }
+          }
+        }
+      }
+    },
+    ""/api/Persons/upload"": {
+      ""post"": {
+        ""operationId"": ""Persons_Upload"",
+        ""consumes"": [
+          ""application/octet-stream"",
+          ""multipart/form-data""
+        ],
+        ""parameters"": [
+          {
+            ""name"": ""data"",
+            ""in"": ""body"",
+            ""schema"": {
+              ""type"": ""string"",
+              ""format"": ""binary"",
+              ""x-nullable"": true
+            },
+            ""x-nullable"": true
+          }
+        ],
+        ""responses"": {
+          ""200"": {
+            ""x-nullable"": true,
+            ""description"": """",
+            ""schema"": {
+              ""type"": ""string"",
+              ""format"": ""byte""
+            }
+          }
+        }
+      }
+    }
+  }
+}";
+
+            var document = await OpenApiDocument.FromJsonAsync(json);
+
+            // Act
+            var settings = new CSharpClientGeneratorSettings { ClassName = "MyClass" };
+            var generator = new CSharpClientGenerator(document, settings);
+            var code = generator.GenerateFile();
+
+            // Assert
+            Assert.DoesNotContain("System.Convert.ChangeType(response", code);
+            Assert.Contains("await ReadObjectResponseAsync<bool>", code);
+            Assert.Contains("await ReadObjectResponseAsync<double?>", code);
+            Assert.Contains("await ReadObjectResponseAsync<string>", code);
+            Assert.Contains("await ReadObjectResponseAsync<byte[]>", code);
         }
     }
 }

--- a/src/NSwag.CodeGeneration.CSharp.Tests/ResponseTests.cs
+++ b/src/NSwag.CodeGeneration.CSharp.Tests/ResponseTests.cs
@@ -177,5 +177,177 @@ namespace NSwag.CodeGeneration.CSharp.Tests
             Assert.Contains("System.Threading.Tasks.Task<System.Collections.Generic.ICollection<BusinessException>>", code);
             Assert.Contains("class BusinessException", code);
         }
+
+        [Fact]
+        public async Task When_responses_produce_multiple_types()
+        {
+            string json = @"{
+  ""openapi"": ""3.0.1"",
+  ""info"": {
+    ""title"": ""WebApplication"",
+    ""version"": ""1.0""
+  },
+  ""servers"": [
+    {
+      ""url"": ""/WebApplication""
+    }
+  ],
+  ""paths"": {
+    ""/Account/Authenticate"": {
+      ""post"": {
+        ""tags"": [
+          ""Account""
+        ],
+        ""requestBody"": {
+          ""content"": {
+            ""application/json"": {
+              ""schema"": {
+                ""$ref"": ""#/components/schemas/AuthenticationRequest""
+              }
+            }
+          }
+        },
+        ""responses"": {
+          ""200"": {
+            ""description"": ""Success"",
+            ""content"": {
+              ""text/plain"": {
+                ""schema"": {
+                  ""type"": ""string""
+                }
+              }
+            }
+          },
+          ""400"": {
+            ""description"": ""Bad Request"",
+            ""content"": {
+              ""application/problem+json"": {
+                ""schema"": {
+                  ""$ref"": ""#/components/schemas/ValidationProblemDetails""
+                }
+              }
+            }
+          },
+          ""401"": {
+            ""description"": ""Unauthorized"",
+            ""content"": {
+              ""application/problem+json"": {
+                ""schema"": {
+                  ""$ref"": ""#/components/schemas/ProblemDetails""
+                }
+              }
+            }
+          },
+          ""415"": {
+            ""description"": ""Unsupported Media Type"",
+            ""content"": {
+              ""application/problem+json"": {
+                ""schema"": {
+                  ""$ref"": ""#/components/schemas/ProblemDetails""
+                }
+              }
+            }
+          },
+          ""500"": {
+            ""description"": ""Internal Server Error"",
+            ""content"": {
+              ""application/problem+json"": {
+                ""schema"": {
+                  ""$ref"": ""#/components/schemas/ProblemDetails""
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  ""components"": {
+    ""schemas"": {
+      ""AuthenticationRequest"": {
+        ""type"": ""object"",
+        ""additionalProperties"": false
+      },
+      ""ValidationProblemDetails"": {
+        ""type"": ""object"",
+        ""properties"": {
+          ""type"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""title"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""status"": {
+            ""type"": ""integer"",
+            ""format"": ""int32"",
+            ""nullable"": true
+          },
+          ""detail"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""instance"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""errors"": {
+            ""type"": ""object"",
+            ""additionalProperties"": {
+              ""type"": ""array"",
+              ""items"": {
+                ""type"": ""string""
+              }
+            },
+            ""nullable"": true,
+            ""readOnly"": true
+          }
+        },
+        ""additionalProperties"": { }
+      },
+      ""ProblemDetails"": {
+        ""type"": ""object"",
+        ""properties"": {
+          ""type"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""title"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""status"": {
+            ""type"": ""integer"",
+            ""format"": ""int32"",
+            ""nullable"": true
+          },
+          ""detail"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          },
+          ""instance"": {
+            ""type"": ""string"",
+            ""nullable"": true
+          }
+        },
+        ""additionalProperties"": { }
+      }
+    }
+  }
+}";
+
+            var document = await OpenApiDocument.FromJsonAsync(json);
+
+            // Act
+            var settings = new CSharpClientGeneratorSettings { ClassName = "MyClass" };
+            var generator = new CSharpClientGenerator(document, settings);
+            var code = generator.GenerateFile();
+
+            // Act
+            Assert.Contains("var result_ = (string)System.Convert.ChangeType(responseData_, typeof(string));", code);
+            Assert.Contains("await ReadObjectResponseAsync<ValidationProblemDetails>", code);
+            Assert.Contains("await ReadObjectResponseAsync<ProblemDetails>", code);
+        }
     }
 }

--- a/src/NSwag.CodeGeneration.CSharp/Models/CSharpOperationModel.cs
+++ b/src/NSwag.CodeGeneration.CSharp/Models/CSharpOperationModel.cs
@@ -73,7 +73,10 @@ namespace NSwag.CodeGeneration.CSharp.Models
 
             Parameters = parameters
                 .Select(parameter =>
-                    new CSharpParameterModel(parameter.Name, GetParameterVariableName(parameter, _operation.Parameters),
+                    new CSharpParameterModel(
+                        parameter.Name,
+                        GetParameterVariableName(parameter, _operation.Parameters),
+                        GetParameterVariableIdentifier(parameter, _operation.Parameters),
                         ResolveParameterType(parameter), parameter, parameters,
                         _settings.CodeGeneratorSettings,
                         _generator,
@@ -225,6 +228,15 @@ namespace NSwag.CodeGeneration.CSharp.Models
         {
             var name = base.GetParameterVariableName(parameter, allParameters);
             return ReservedKeywords.Contains(name) ? "@" + name : name;
+        }
+
+        /// <summary>Gets the identifier of the parameter variable.</summary>
+        /// <param name="parameter">The parameter.</param>
+        /// <param name="allParameters">All parameters.</param>
+        /// <returns>The parameter variable identifier.</returns>
+        protected string GetParameterVariableIdentifier(OpenApiParameter parameter, IEnumerable<OpenApiParameter> allParameters)
+        {
+            return base.GetParameterVariableName(parameter, allParameters);
         }
 
         /// <summary>Resolves the type of the parameter.</summary>

--- a/src/NSwag.CodeGeneration.CSharp/Models/CSharpParameterModel.cs
+++ b/src/NSwag.CodeGeneration.CSharp/Models/CSharpParameterModel.cs
@@ -20,6 +20,7 @@ namespace NSwag.CodeGeneration.CSharp.Models
         /// <summary>Initializes a new instance of the <see cref="CSharpParameterModel" /> class.</summary>
         /// <param name="parameterName">Name of the parameter.</param>
         /// <param name="variableName">Name of the variable.</param>
+        /// <param name="variableIdentifier">Identifier of the variable.</param>
         /// <param name="typeName">The type name.</param>
         /// <param name="parameter">The parameter.</param>
         /// <param name="allParameters">All parameters.</param>
@@ -29,6 +30,7 @@ namespace NSwag.CodeGeneration.CSharp.Models
         public CSharpParameterModel(
             string parameterName,
             string variableName,
+            string variableIdentifier,
             string typeName,
             OpenApiParameter parameter,
             IList<OpenApiParameter> allParameters,
@@ -37,6 +39,7 @@ namespace NSwag.CodeGeneration.CSharp.Models
             TypeResolverBase typeResolver)
             : base(parameterName, variableName, typeName, parameter, allParameters, settings, generator, typeResolver)
         {
+            this.VariableIdentifier = variableIdentifier;
         }
 
         /// <summary>Gets a value indicating whether the type is a Nullable&lt;&gt;.</summary>
@@ -54,5 +57,8 @@ namespace NSwag.CodeGeneration.CSharp.Models
             Schema.AllowAdditionalProperties &&
             !IsDictionary &&
             Type != "object";
+
+        /// <summary>Gets the unescaped variable name.</summary>
+        public string VariableIdentifier { get; }
     }
 }

--- a/src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.ProcessResponse.liquid
+++ b/src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.ProcessResponse.liquid
@@ -19,7 +19,7 @@ return new {{ ResponseClass }}<{{ operation.UnwrappedResultType }}>(status_, hea
 return result_;
 {%             endif -%}
 {%         else -%}
-throw new {{ ExceptionClass }}<{{ response.Type }}>("{{ response.ExceptionDescription }}", status_, objectResponse_.Text, headers_, result_, null);
+throw new {{ ExceptionClass }}<{{ response.Type }}>("{{ response.ExceptionDescription }}", status_, responseData_, headers_, result_, null);
 {%         endif -%}
 {%     else -%}
 var objectResponse_ = await ReadObjectResponseAsync<{{ response.Type }}>(response_, headers_, cancellationToken).ConfigureAwait(false);

--- a/src/NSwag.CodeGeneration.CSharp/Templates/Client.Method.Documentation.liquid
+++ b/src/NSwag.CodeGeneration.CSharp/Templates/Client.Method.Documentation.liquid
@@ -3,7 +3,7 @@
 {% endif -%}
 {% for parameter in operation.Parameters -%}
 {%     if parameter.HasDescription -%}
-/// <param name="{{ parameter.VariableName }}">{{ parameter.Description | csharpdocs }}</param>
+/// <param name="{{ parameter.VariableIdentifier }}">{{ parameter.Description | csharpdocs }}</param>
 {%     endif -%}
 {% endfor -%}
 {% if operation.HasResultDescription -%}

--- a/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
+++ b/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
@@ -89,13 +89,8 @@ namespace NSwag.CodeGeneration.Models
         {
             get
             {
-                JsonObjectType primitive = JsonObjectType.String | JsonObjectType.Integer | JsonObjectType.Number | JsonObjectType.Boolean;
-
                 return !_response.Content.ContainsKey("application/json") &&
-                       (_response.Content.ContainsKey("text/plain") || _operationModel.Produces == "text/plain")
-                       && (ActualResponseSchema != null) &&
-                            ((ActualResponseSchema.Type & primitive) == ActualResponseSchema.Type) &&
-                            ((ActualResponseSchema.Type | primitive) == primitive);
+                       _response.Content.ContainsKey("text/plain");
             }
         }
 

--- a/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
+++ b/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
@@ -94,8 +94,8 @@ namespace NSwag.CodeGeneration.Models
                 return !_response.Content.ContainsKey("application/json") &&
                        (_response.Content.ContainsKey("text/plain") || _operationModel.Produces == "text/plain")
                        && (ActualResponseSchema != null) &&
-                            ((ActualResponseSchema.ActualTypeSchema.Type & primitive) == ActualResponseSchema.ActualTypeSchema.Type) &&
-                            ((ActualResponseSchema.ActualTypeSchema.Type | primitive) == primitive);
+                            ((ActualResponseSchema.Type & primitive) == ActualResponseSchema.Type) &&
+                            ((ActualResponseSchema.Type | primitive) == primitive);
             }
         }
 

--- a/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
+++ b/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
@@ -85,9 +85,16 @@ namespace NSwag.CodeGeneration.Models
         }
 
         /// <summary>Gets a value indicating whether the response requires a text/plain content.</summary>
-        public bool IsPlainText =>
-            !_response.Content.ContainsKey("application/json") &&
-            (_response.Content.ContainsKey("text/plain") || _operationModel.Produces == "text/plain");
+        public bool IsPlainText
+        {
+            get
+            {
+                JsonObjectType primitive = JsonObjectType.String | JsonObjectType.Integer | JsonObjectType.Number | JsonObjectType.Boolean;
+                return ActualResponseSchema != null &&
+                    (ActualResponseSchema.ActualTypeSchema.Type & primitive) == ActualResponseSchema.ActualTypeSchema.Type &&
+                    (ActualResponseSchema.ActualTypeSchema.Type | primitive) == primitive;
+            }
+        }
 
         /// <summary>Gets a value indicating whether this is a file response.</summary>
         public bool IsFile => IsSuccess && _response.IsBinary(_operation);

--- a/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
+++ b/src/NSwag.CodeGeneration/Models/ResponseModelBase.cs
@@ -90,9 +90,12 @@ namespace NSwag.CodeGeneration.Models
             get
             {
                 JsonObjectType primitive = JsonObjectType.String | JsonObjectType.Integer | JsonObjectType.Number | JsonObjectType.Boolean;
-                return ActualResponseSchema != null &&
-                    (ActualResponseSchema.ActualTypeSchema.Type & primitive) == ActualResponseSchema.ActualTypeSchema.Type &&
-                    (ActualResponseSchema.ActualTypeSchema.Type | primitive) == primitive;
+
+                return !_response.Content.ContainsKey("application/json") &&
+                       (_response.Content.ContainsKey("text/plain") || _operationModel.Produces == "text/plain")
+                       && (ActualResponseSchema != null) &&
+                            ((ActualResponseSchema.ActualTypeSchema.Type & primitive) == ActualResponseSchema.ActualTypeSchema.Type) &&
+                            ((ActualResponseSchema.ActualTypeSchema.Type | primitive) == primitive);
             }
         }
 

--- a/src/NSwag.Integration.ClientPCL.Tests/NSwag.Integration.ClientPCL.Tests.csproj
+++ b/src/NSwag.Integration.ClientPCL.Tests/NSwag.Integration.ClientPCL.Tests.csproj
@@ -17,6 +17,7 @@
     <IsCodedUITest>False</IsCodedUITest>
     <TestProjectType>UnitTest</TestProjectType>
     <TargetFrameworkProfile />
+    <RuntimeIdentifiers>win</RuntimeIdentifiers>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
     <DebugSymbols>true</DebugSymbols>
@@ -80,6 +81,14 @@
       <Project>{ECD41518-A92C-4706-8E74-9707AFA554B7}</Project>
       <Name>NSwag.Integration.ClientPCL</Name>
     </ProjectReference>
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json">
+      <Version>10.0.2</Version>
+    </PackageReference>
+    <PackageReference Include="System.ComponentModel.Annotations">
+      <Version>4.1.0</Version>
+    </PackageReference>
   </ItemGroup>
   <Choose>
     <When Condition="'$(VisualStudioVersion)' == '10.0' And '$(IsCodedUITest)' == 'True'">

